core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java

```diff
@@ -453,11 +455,13 @@ public boolean ownsData() {
 
    private void fetchClusterListeners(CacheTopology cacheTopology) {
       if (configuration.clustering().cacheMode().isDistributed() || configuration.clustering().cacheMode().isScattered()) {
-         Collection<DistributedCallable> callables = getClusterListeners(cacheTopology);
-         for (DistributedCallable callable : callables) {
-            callable.setEnvironment(cache.wired(), null);
+         Collection<ClusterListenerReplicateCallable<Object, Object>> callables = getClusterListeners(cacheTopology);
+         Cache<Object, Object> cache = this.cache.wired();
+         for (ClusterListenerReplicateCallable<Object, Object> callable : callables) {
             try {
-               callable.call();
+               // TODO: need security check?
+               // We have to invoke a separate method as we can't retrieve the cache as it is still starting
+               callable.accept(cache.getCacheManager(), cache);
```

Comment:
TBH I always found it odd that we'd get back a `Callable` instead of a POJO with the information we need to register the listener ourselves. If we did the registration ourselves we'd just call a method on the `CacheNotifier`, we wouldn't need the cache.

Reply:
There are a couple parts to this reason. The first is that because this is invoked remotely it was using `DistributedExecutor` to do that instead of having to add a whole new command. This therefore requires the `Callable` approach. The second item is then the `StateConsumerImpl` doesn't need to know how to register it properly, if you notice there is quite a bit of code to make sure the listener is only registered once - which I wouldn't think the `StateConsumerImpl` would want in it. Also the `Callable` has changed a few times already - which isolates how many class changes are required.

Reply:
The callable is the response from the remote node, so it could have been any serializable POJO. And `StateConsumerImpl` can delegate the registration to `ClusterCacheNotifier`, just like `StateProviderImpl` uses `retrieveClusterListenerCallablesToInstall()`.

The fact that registration changes from version to version is actually one more reason to avoid the Callables, because it's much harder to ensure that changes from one version to another are compatible for the future rolling upgrades.

Reply:
Well in that case the POJO or Callable would be a different class or have different members, complicating the whole thing :(

It sounds like we may want to change this as you mentioned, but probably in the scope of upgrade JIRA so it is done properly.

Reply:
Agree, it's out of scope, I just added the comment because I saw the TODO